src/Workspaces/Core/MSBuild.BuildHost/IRemoteProjectFile.cs

```diff
@@ -25,4 +25,15 @@ internal interface IRemoteProjectFile : IDisposable
 {
     Task<ImmutableArray<ProjectFileInfo>> GetProjectFileInfosAsync(CancellationToken cancellationToken);
     Task<ImmutableArray<DiagnosticLogItem>> GetDiagnosticLogItemsAsync(CancellationToken cancellationToken);
+
+    Task<string> GetDocumentExtensionAsync(SourceCodeKind sourceCodeKind, CancellationToken cancellationToken);
+    Task AddDocumentAsync(string filePath, string? logicalPath, CancellationToken cancellationToken);
```

Comment:
for my underatnding, what's the diff between a filepath and a logicalpath?

Reply:
File path is physical file on disk. In the linked file case, the file appears in solution explorer and can be added in a subdirectory. For example, consider places where we might link in the workspaces layer to some files in the compiler layer, but in the process we then say "oh put it in the CompilerUtilities subfolder". File path is the physical path on disk, logical path is "its in CompilerUtilities".